crates/bevy_core_widgets/src/callback.rs

```diff
@@ -27,15 +29,119 @@ use bevy_ecs::world::{DeferredWorld, World};
 /// // Later, when we want to execute the callback:
 /// app.world_mut().commands().notify(&callback);
 /// ```
-#[derive(Default, Debug)]
+#[derive(Debug)]
 pub enum Callback<I: SystemInput = ()> {
     /// Invoke a one-shot system
     System(SystemId<I>),
     /// Ignore this notification
+    Ignore,
+}
+
+impl<I: SystemInput> Copy for Callback<I> {}
+impl<I: SystemInput> Clone for Callback<I> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::System(arg0) => Self::System(arg0.clone()),
+            Self::Ignore => Self::Ignore,
+        }
+    }
+}
+
+impl<In: SystemInput + 'static> GetTemplate for Callback<In> {
+    type Template = CallbackTemplate<In>;
+}
+
+#[derive(Default)]
+pub enum CallbackTemplate<In: SystemInput = ()> {
+    System(Box<dyn RegisterSystem<In>>),
+    SystemId(SystemId<In>),
     #[default]
     Ignore,
 }
 
+impl<In: SystemInput + 'static> CallbackTemplate<In> {
+    pub fn clone(&self) -> CallbackTemplate<In> {
+        match self {
+            CallbackTemplate::System(register_system) => {
+                CallbackTemplate::System(register_system.box_clone())
+            }
+            CallbackTemplate::SystemId(system_id) => CallbackTemplate::SystemId(*system_id),
+            CallbackTemplate::Ignore => CallbackTemplate::Ignore,
+        }
+    }
+}
+
+pub trait RegisterSystem<In: SystemInput>: Send + Sync + 'static {
+    fn register_system(&mut self, world: &mut World) -> SystemId<In>;
+    fn box_clone(&self) -> Box<dyn RegisterSystem<In>>;
+}
+
+pub struct IntoWrapper<I, In, Marker> {
+    into_system: Option<I>,
+    marker: PhantomData<fn() -> (In, Marker)>,
+}
+
+pub fn callback<
+    I: IntoSystem<In, (), Marker> + Send + Sync + Clone + 'static,
+    In: SystemInput + 'static,
+    Marker: 'static,
+>(
+    system: I,
+) -> CallbackTemplate<In> {
+    CallbackTemplate::from(IntoWrapper {
+        into_system: Some(system),
+        marker: PhantomData,
+    })
+}
+
+impl<
+        I: IntoSystem<In, (), Marker> + Clone + Send + Sync + 'static,
+        In: SystemInput + 'static,
+        Marker: 'static,
+    > RegisterSystem<In> for IntoWrapper<I, In, Marker>
+{
+    fn register_system(&mut self, world: &mut World) -> SystemId<In> {
+        world.register_system(self.into_system.take().unwrap())
```

Comment:
Does this leak? We need a way to unregister the callback when the owner is despawned.